### DISCUSSION TITLE
unified-engine: Respect directory hierarchy when outputting to a different directory

### DISCUSSION THREAD

**Discussion by NickSto** [unifiedjs/unified#138] [Q&A] [answered]
I'm trying to wrap my head around [unified-engine](https://github.com/unifiedjs/unified-engine), and whether it can be used for what I'd like to do.

I'd like to take a directory full of many Markdown files in many subdirectories and write the output to a separate directory, maintaining the same directory tree, directory names, and filenames. Basically I'd like a copy of the original directory, but with certain modifications to the Markdown content.

I found that I can direct the output to a different directory using [options.output](https://github.com/unifiedjs/unified-engine/blob/main/doc/options.md#optionsoutput), but when I do that the directory structure gets flattened and all files end up in the same directory. Is there a way to get it to maintain the structure?

Also, I was curious about whether there's a way to specify the output file extension? If I want to convert `.md` to `.html` that'd be useful.

**Comment by wooorm** [comment, marked as answer]
> Q1: keep directory structure intact

Nope, not supported.
I think Gulp does some interesting things here (which was briefly [discussed before here](https://github.com/vfile/ideas/issues/4)), but how should it work?

Some examples, with the following input files:

```
input/a/readme.md
input/b/readme.md
readme.md
```

```sh
$ ./node_modules/.bin/remark -u html -o output/ input/
$ ./node_modules/.bin/remark -u html -o output/ input/ readme.md
$ ./node_modules/.bin/remark -u html -o output/ input/{a,b}
$ ./node_modules/.bin/remark -u html -o output/ **/*.md
```

What should the structure of `output/` be?

> Q2: Extensions

Plugins can change `file`s. In the above examples, `remark-html` is used, which [sets `extnames` of  `.html`](https://github.com/remarkjs/remark-html/blob/357ca5860c04cb6bf4ef0caeda511ecc05aca274/index.js#L22-L24), and hence the written files will have `.html` extensions

**Reply by NickSto** [reply]
Cool, thanks for clarifying!

In your example I'm not sure how the `remark` command works but if the `unified-engine` setup is
```Javascript
engine(
  {
    processor: processor,
    files: ['input'],
    extensions: ['md'],
    output: 'output',
    flatten: false,
      ...
```
(if `flatten: false` was an option that made it keep the directory structure)

And the contents of `input` are
```
input/a/readme.md
input/b/readme.md
input/readme.md
```
Then afterward, the contents of `output` should be
```
output/a/readme.md
output/b/readme.md
output/readme.md
```

**Reply by wooorm** [reply]
`remark-cli` (https://github.com/remarkjs/remark/blob/main/packages/remark-cli/cli.js) is a tiny wrapper around `unified-args`, which in turn is a tiny wrapper around `unified-engine`.

It looks like your example would replace the “common ancestor” of all resolved files.
What, in your case, should happen for?

```
input/a/b/readme.md
input/a/c/readme.md
```

Then it sounds like both `input/` and `a/` would be stripped, which seems weird (and hence why that “glob base” that gulp uses seems smarter, although not perfect, to me?)

Note: or, is `input` the base because it’s given in the unified engine args?

**Reply by NickSto** [reply]
Yeah, sorry, maybe I'm not being clear. The situation I'm describing is where I have a directory `input`, full of Markdown files and subdirectories, as shown, and an empty directory `output`. I configure `unified-engine` like I showed, then execute the script.

Afterward, `output` should have the same contents as `input`, except the Markdown file contents are modified by the processor.

Maybe the `unified-engine` configuration I showed is wrong? I only just started experimenting with it.

**Reply by wooorm** [reply]
Nah, I'm assuming you're posting pseudo code, I'm doing that too, but I guess my Q is: how can the code know 'input/' is the base directory to skip?

**Reply by NickSto** [reply]
Oh okay, I think I see what you're asking. Yeah, I'm not sure how to allow giving multiple paths to the `files` option when there's only one path given to `output`. I've only been thinking of the situation where you want to copy the contents of a single directory into another, single directory.

Maybe when it's operating in this mode (i.e. `flatten: false`), you only allow one path in the `files` array. Or you give the single path to a new option.

Either way, `unified-engine` gets one input path, and assumes that's the root directory whose contents you want to duplicate into the output directory.

**Reply by wooorm** [reply]
I’d rather not add a new option, that is so specific as to only allow a single input file. I’d have a hard time documenting and explaining it to users in a sensible way.

Assuming a path to a directory is given, the path to that directory could be stripped, and the rest remain. Same for a glob, like Gulp, the non-magic part starting it off could be stripped as well.
Still, it seems complex and needs some fiddling to test it out and see if it works

**Reply by janaka** [reply]
Hi, 
I think what I'm needing to do is the same as @NickSto. I have a folder with a bunch of sub folders which contain MD files. I need to process each file and output as copy elsewhere maintaining the relative folder structure. So if the source was in relative for `/a` then the output is created in folder`/a` relative to `output`, folder created if doesn't exist.  To add a bit more colour, my MD files are docs with code blocks in language A. I'm creating version of the same docs for language B and C.

original source docs
```
/orig/a/doc1.md
/orig/b/doc2.md
/orig/index.md
```

converted version for java
```
/java/a/doc1.md
/java/b/doc2.md
/java/index.md
```

```
files: ['/orig/'].
output: '/java/'
```

As mentioned above, it doesn't do that, which I was a little surprised given folder are integrated on the input side.

Given the state of things my only option is to not use unifed-engine and instead write the input/output code myself right?

**Reply by wooorm** [reply]
You can also call the cli several times?

This is still rather vague about how it would work, there are many edge cases to think through before this can be worked on
